### ISSUE TITLE
PayPal Express should handle errors on subscription creation

### ISSUE THREAD

**Issue by mircobabini** [strangerstudios/paid-memberships-pro#1590]
**Spoiler: probably found a solution 🎉**
tldr; Found that there is a difference between a CreateRecurringPaymentProfile success and an insuccess. In the first case, there is [PROFILESTATUS] => ActiveProfile; when it fails instead, even if ACK=SUCCESS, PROFILESTATUS is completely missing. [Fix idea](https://github.com/strangerstudios/paid-memberships-pro/compare/dev...mircobabini:patch-14).

-----

**Describe the bug**
When the users gets back from the gateway we know if the initial payment was successful or not. This is true for some gateways, but it's not always true for PayPal Express. Sometimes PayPal Express returns not enough informations to say "yes, the order was really successful". PMPro handles this uncertainty with this flow:
1. order in status token is created before the gateway call
1. order status becomes review and suddenly success, when gateway returns to the site
1. order status becomes error when on the ipn call reporting "Initial Payment = Failed"

**Consequences**

Confusion: the customer receive a success email, confirming its membership. He logs in, but after a minute or two he is not able to continue on site; he is redirected to the subscription page once again... not clear at all.

Invoices add-on: legal invoices got generated when order status is set to success, and should be invalidated when the order status becomes error. This is not possible, because every legal document emitted is emitted; can't just be invalidated. In fact, it should not be emitted at all. As already discussed, this is probably acceptable in US... but it's causing issues in EU, expecially in Italy because of e-invoice which must be precise, fast and essential to be considered valid and to avoid fines.

Gifts add-on: gift coupon code are sent when the order status is set to success, but they should be revoken after 2 minutes. This is impossible and confusing.

**Ideas**
1. The review status should be kept for a while; we can wait for the coming soon ipn response, which will change the order status to success or error. Meanwhile, we can set up an option to decide if the customer should be able to use the membership area or if he should see an informative message of the idle moment. Please note that there are possibilities for the ipn response to never come. These are rare cases I was able to solve using a cron job; every night, I'm checking the orders in review status, older than 24 hours (the day before). If there are some, I call the getSubscriptionStatus gateway feature and I'm 100% able to see if the initial payment was a success or an error. Tricky, but at least I was able to remove the orders in review status, which is not very informative later on.
1. We can use the getSubscriptionStatus gateway feature to better understand what's going on.
1. We can investigate and see if there is something helpful to avoid issues, with no additional requests.  **This is what I've tried to do in the following analysis.**

**The flow**
### Status = token

Thanks to the *pmpro_added_order* action I was able to detect the exact moment the MemberOrder was created, inserted into the database in status=token. This happens during the checkout page submit (submit-checkout=1), then the user is redirected to PayPal Express.

### Status = review

This is happening when the user get back from PayPal Express. The current request is similar to the checkout page with some additional params: &review={PMPRO_ORDER_CODE}&token={PAYPAL_TOKEN_EC-XXXXXXXXXXXXX}&PayerID={PAYPAL_PAYER_IDENTIFIER}.

Thanks to getExpressCheckoutDetails() we know from PayPal API[*GetExpressCheckoutDetails*] if the checkout was ACK=SUCCESS (causing order status = review) or ACK=SUCCESSWITHWARNING (causing order status = error). [https://github.com/mircobabini/paid-memberships-pro/blob/60ab16975ad75cdb78340512a32390f1468dc1aa/classes/gateways/class.pmprogateway_paypalexpress.php#L328](https://github.com/mircobabini/paid-memberships-pro/blob/60ab16975ad75cdb78340512a32390f1468dc1aa/classes/gateways/class.pmprogateway_paypalexpress.php#L328)

### Status = success

During the same routine as above, if PayPal Express is set to skip confirmation, $order→confirm() is called; this one is responsible of →subscribe(), running the PayPal API[*CreateRecurringPaymentsProfile*] to set up the recurrent payment: ACK=SUCCESS (causing order status = success) or ACK=SUCCESSWITHWARNING (causing order status = error). [https://github.com/mircobabini/paid-memberships-pro/blob/60ab16975ad75cdb78340512a32390f1468dc1aa/classes/gateways/class.pmprogateway_paypalexpress.php#L383](https://github.com/mircobabini/paid-memberships-pro/blob/60ab16975ad75cdb78340512a32390f1468dc1aa/classes/gateways/class.pmprogateway_paypalexpress.php#L383)

🔜: this is where we should fix the payment_transaction_id: [https://github.com/mircobabini/paid-memberships-pro/blob/60ab16975ad75cdb78340512a32390f1468dc1aa/classes/gateways/class.pmprogateway_paypalexpress.php#L747-L748](https://github.com/mircobabini/paid-memberships-pro/blob/60ab16975ad75cdb78340512a32390f1468dc1aa/classes/gateways/class.pmprogateway_paypalexpress.php#L747-L748). Must be collected somehow. Weird thing is that PayPal docs says we should have a TRANSACTIONID but it's missing. Let's see if PayPal Checkout can do something more. Related issue: #1077.

💡 Found that there is a difference between a CreateRecurringPaymentProfile success and an insuccess. In the first case, there is [PROFILESTATUS] => ActiveProfile; when it fails instead, even if ACK=SUCCESS, PROFILESTATUS is completely missing. This can be used to avoid setting the order status=success here: [https://github.com/strangerstudios/paid-memberships-pro/blob/d0a252dfa967dca080f95cedacb1982410ff6847/classes/gateways/class.pmprogateway_paypalexpress.php#L745-L753](https://github.com/strangerstudios/paid-memberships-pro/blob/d0a252dfa967dca080f95cedacb1982410ff6847/classes/gateways/class.pmprogateway_paypalexpress.php#L745-L753)

[This is the proof of concept fix](https://github.com/strangerstudios/paid-memberships-pro/compare/dev...mircobabini:patch-14). Must be completed, of course subscribe() must then return false at least... and must check the flow.

### Status = error

After a while, the ipn response comes from PayPal with txn_type = recurring_payment_profile_cancel, initial_payment_status = Failed. This was already subject of a recent (2020) fix about setting order status to error, instead of cancel, to avoid misunderstanding and not considering errored orders in analytics totals.

---

Personal note: this was debugged on order_id = 91424 (first data).
Then the missing PROFILESTATUS field is confirmed in dozens of orders on March 6, 2021.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only Paid Memberships Pro plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Memberlite](https://wordpress.org/themes/memberlite/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by mircobabini** [comment]
Finally, just logged a full order with error flow.

[06-Mar-2021 17:21:21 UTC] ---------- MemberOrder::process()
[06-Mar-2021 17:21:22 UTC] 01FAF39A09 added (token)
[06-Mar-2021 17:21:42 UTC] 01FAF39A09 updated (review)
[06-Mar-2021 17:21:42 UTC] 01FAF39A09 MemberOrder::confirm()
[06-Mar-2021 17:21:44 UTC] 01FAF39A09 MemberOrder::subscribe(): ACK=SUCCESS
[06-Mar-2021 17:21:44 UTC] 01FAF39A09 PROFILESTATUS is missing; should set ->status = error! ⚠️
[06-Mar-2021 17:21:45 UTC] 01FAF39A09 updated (success) ⚪️
[06-Mar-2021 17:21:46 UTC] 01FAF39A09 updated (success) ⚪️
[06-Mar-2021 17:21:46 UTC] 01FAF39A09 checkout done (success) for user_id=2234
[06-Mar-2021 17:22:39 UTC] 01FAF39A09 updateStatus from success to cancelled (🔵)
[06-Mar-2021 17:22:40 UTC] 01FAF39A09 updateStatus from cancelled to error (🔵)

For unclear logged rows (⚪️) and ipn logged rows (🔵) I have collected some more data in email. Will check it out tomorrow.

Please note this one taken about 55 seconds before receive the IPN Response and change status from success to error. Meanwhile, the user was inside its member area... then suddenly got thrown out.

**Comment by mircobabini** [comment]
Just collected an evidence of an order which remained in _status=review_ until midnight; then my cron job switched it to error. Otherwise, it would be a _review_ forever. Note to myself: order code = E2D43FD7FB. This is the flow:

[07-Mar-2021 06:41:20 UTC] E2D43FD7FB added (token)
[07-Mar-2021 06:42:22 UTC] E2D43FD7FB updated (review)
[07-Mar-2021 06:42:23 UTC] E2D43FD7FB MemberOrder::confirm()
[07-Mar-2021 06:42:25 UTC] E2D43FD7FB MemberOrder::subscribe(): ACK=SUCCESS
[07-Mar-2021 06:42:25 UTC] E2D43FD7FB PROFILESTATUS is missing; should set ->status = error! ⚠️
[07-Mar-2021 06:42:26 UTC] E2D43FD7FB updated (success) ⚪️
[07-Mar-2021 06:42:26 UTC] E2D43FD7FB updated (success) ⚪️
[07-Mar-2021 06:42:27 UTC] E2D43FD7FB checkout done (success) for user_id=8019
After just 1 second, seems like the page reloaded or something. Because of this, preheaders/checkout loaded once again, but this time of course the ACK would be Failure... and since we are setting the order status to review once again, it will stay in this status forever.
[07-Mar-2021 06:42:28 UTC] E2D43FD7FB updated (review)
[07-Mar-2021 06:42:29 UTC] E2D43FD7FB MemberOrder::confirm()
[07-Mar-2021 06:42:30 UTC] E2D43FD7FB MemberOrder::subscribe(): ACK=FAILURE
[07-Mar-2021 06:43:13 UTC] E2D43FD7FB MemberOrder::confirm()
[07-Mar-2021 06:43:14 UTC] E2D43FD7FB MemberOrder::subscribe(): ACK=FAILURE
...until, at midnight, my cron job got the order in status review, checked the status with the PayPal API and switched to error.
[08-Mar-2021 01:10:06 UTC] E2D43FD7FB updateStatus from review to error

**Considerations:**
1. Could be set to error since PROFILESTATUS was missing.
2. By the way, we should limit the status change somehow. An order in status=token can be changed to status=review; but an order in status=success|error|refunded|cancelled should never become status=review.

The idea is to avoid the double preheaders/checkout processing if the order `status != token`, to avoid double processing.

**Comment by mircobabini** [comment]
When using PayPal CreateRecurringPaymentsProfile the PROFILESTATUS can be:
1. missing => from experience, it's an error 100%
2. =ActiveProfile => first payment collected, great.
3. =PendingProfile => first payment not collected, but PayPal will try to collect it.

Since today we were not checking the PROFILESTATUS field, so we were considering it always valid (even if missing). After investigation above and looking at some implementations from others (i.e. Laravel) we can confirm the three conditions above.

PendingProfile is unclear; from the official doc:
<img width="858" alt="Screenshot 2021-03-09 at 10 59 07" src="https://user-images.githubusercontent.com/636911/110454058-1db23e80-80c7-11eb-925c-035519552c13.png">

This means "not collected yet, but probably will be". This is why many others are considering it as processed and the PayPal IPN process will do the rest, if something would be wrong later on. Great. There's an additional param to control the PayPal behaviour for PendingProfile situations:
<img width="861" alt="Screenshot 2021-03-09 at 10 59 25" src="https://user-images.githubusercontent.com/636911/110454397-797cc780-80c7-11eb-82df-438d9b81afdd.png">

Since we are also setting `MAXFAILEDPAYMENTS=1`, we want to block the user as soon as we know he was unable to complete the payment. So it's ok to keep the `FailedInitialAmountAction` unset, which means `CancelOnFailure`. In the worst case scenario, the member would have been on the platform just for a while.

While checking some other implementations I found this interesting thread: https://stackoverflow.com/q/25145788/1160173
Seems like the `INITAMT` when performing the CreateRecurringPaymentsProfile is the cause. As far as I can see, WooCommerce Subscriptions is not using `INITAMT`. Instead, they are doing exactly how explained in that StackOverflow, when performing the `SetExpressCheckout`. 

References:
1. [WCS_PayPal_Reference_Transaction_API_Request::set_express_checkout](https://github.com/wp-premium/woocommerce-subscriptions/blob/2bbca688afef20b98ca5f5ae4347354458001768/includes/gateways/paypal/includes/class-wcs-paypal-reference-transaction-api-request.php#L68-L127)
1. [WCS_PayPal_Reference_Transaction_API_Request::add_payment_details_parameters](https://github.com/wp-premium/woocommerce-subscriptions/blob/2bbca688afef20b98ca5f5ae4347354458001768/includes/gateways/paypal/includes/class-wcs-paypal-reference-transaction-api-request.php#L304-L310)
1. [WCS_PayPal_Reference_Transaction_API_Request::add_line_item_parameters](https://github.com/wp-premium/woocommerce-subscriptions/blob/2bbca688afef20b98ca5f5ae4347354458001768/includes/gateways/paypal/includes/class-wcs-paypal-reference-transaction-api-request.php#L481-L489)

I will keep this in mind for the future.

**Comment by mircobabini** [comment]
**Heads up:** since we were ignoring the missing of PROFILESTATUS, the preheaders/checkout process used to create the user. Since we are now handling it the right way, now the user is not created anymore. This is the correct behaviour indeed.

Snippet to create the user even if the first payment fails:
https://gist.github.com/mircobabini/9b5bb46ce66b6d2cd612e04cde27b9dc

**Comment by mircobabini** [comment]
**Related PRs status:**
#1598: merged
#1597: merged
~~#1599: rejected (check PR comments)~~
#1600: ready to be merged (3 commits)
